---
title: "About"
output: html_document
---


## OneMap software

The OneMap software was first developed by Gabriel Margarido, Marcelo Mollinari and Antonio Augusto Fraco Garcia and published in 2007(ref). It was one of the first softwares able to build genetic maps for outcrossing species. For this it makes use of two-points estimations for each combinations of markers and a HMM approach which combine informations from every marker in the linkage group to estimate the genetic distances. But, the software was first developted to deal with few and informative genetic markers, as microssatelites, with the modern high-throughput markers OneMap now have to deal with thousands of markers, with low-informativeness and a lot of genotype errors. Without a adaptation of the software it produces inflated maps and takes a lot of time to run.

## Objectives

* Paralelize the genetic distance estimation in OneMap;
* Measure the impact of using different genotype error probabilities in the HMM OneMap approach to estimate the genetic distances.

## Errors workflow

The workflow is written in wdl and it is available in the repository [Cristianetaniguti/errors_workflow](https://github.com/Cristianetaniguti/errors_workflow)

 The errors workflow simulates Illumina reads for simulated bi-parental populations. After the reads simulations, the worflow performs the alignemnt of the reads, the SNP calling, genotype calling and build genetics maps in OneMap software. OneMap now can use genotyping probabilities to estimate the genetic distances, than the genotype probability estimated by each software will be applied to generate different genetic maps. Through the graphics here presented it is possible to measure the effects of the different SNP and genotype calling in the genetic maps.

## Shiny app

This shiny app plots analyses datasets (.rds) generated by the errors_workflow.

Each icon at the left corner of this page point a different dataset and produce interative plots. Each page plots two graphics than comparisions can be made. Also, you can download de graphics using the button at the end of the page.

## Citation

For more informations and for citation access the [paper]().

## Authors

Cristiane Taniguti is currently a phD student at ESALQ/USP Statistical Genetics Lab. Access her personal page [here](http://cristianetaniguti.github.io/), and the page of the lab [here](http://augustogarcia.me/statgen-esalq/).
