# Reads2Map App

Shiny app to present Reads2Map Workflows results in graphics. Here you can see:

## SimulatedReads results

* SNP calling efficiency
* Coverage
* Filters
* Markers type
* Time spent to build the map
* Read depth and genotypes estimated
* Map size
* Phases
* Recombination fraction matrix as hetmaps
* Progeny haplotypes
* Breakpoints count

## EmpicalReads results:

* Coverage
* Filters
* Markers type
* Time spent to build the map
* Read depth and genotypes estimated
* Map size
* Recombination fraction matrix as hetmaps
* Progeny haplotypes
* Breakpoints count

# Quick Start

You can run the shiny app with:

```
library(shiny)
runGitHub("Cristianetaniguti/Reads2MapApp)
```

Acess [here]() to see examples with populus, eucalyptus and acca maps.




