---
title: "About"
output: html_document
---


## OneMap software

The OneMap software was first developed by Gabriel Margarido, Marcelo Mollinari, and Antonio Augusto Franco Garcia and [published in 2007](http://taurus.unicamp.br/bitstream/REPOSIP/81752/1/WOS000248257800002.pdf). It was one of the first software able to build genetic maps for outcrossing species. It estimates a two-point recombination fraction for each combination of markers and applies a multipoint approach to combine information from every marker in the linkage group to estimate the genetic distances. But, the software was first developed to deal with few and informative genetic markers, as microsatellites. With the modern high-throughput markers, OneMap now has to deal with thousands of markers, with low-informativeness and a lot of genotyping errors. Without an adaptation of the software, it produces inflated maps and takes a lot of time to run.

## Reads2Map workflows

Reads2Map workflow uses [WDL language and cromwell](https://gatkforums.broadinstitute.org/wdl/) from [Broad Institute](https://www.broadinstitute.org/) to build linkage maps with user-friendly workflows and optimized usage of memory, CPU, and time. There are two main workflows called SimulatedReads2Map and EmpiricalReads2Map. The first performs population and RADseq FASTQ files simulations for a chromosome, SNP, and genotype calling with five different software and genetic map building with OneMap and GUSMap. EmpiricalReads2Map receives FASTQ files from empirical data and also performs the SNP calling for the entire dataset and genotype calling and map building to a selected chromosome.

The main workflows are composed of sub-workflows that can also be useful for users, they are located in the `tasks` directory. The sub-workflows are composed of tasks that can be found at the end of each `.wdl` file, `utils.wdl`, and `utilsR.wdl`.

[Docker](https://www.docker.com/) images were built to each one of the tasks, then users need only to download the most recent version of cromwell, install its requirements, and docker. If running without sudo permission users can use [singularity](https://singularity.lbl.gov/docs-docker) instead of directly docker. See `.configurations` directory and its [instructions]() to use singularity.

The main goals of Reads2Map workflows are to test software capabilities in simulated scenarios (SimulatedReads2Map) and in empirical data (EmpiricalReads2Map). SimulatedReads2Map also gives a safe method to test new functionalities implemented in OneMap and EmpiricalReads2Map gives a fast and automatic method to select the best software to build linkage maps and answer biological questions for any empirical data.

Output datasets from each main workflow can be easily evaluated in a shiny app called [Reads2MapApp](https://github.com/Cristianetaniguti/Reads2MapApp). You need to submit your output to the app, and the final page gives the possibility to download the OneMap or GUSMap object containing the combination of software and parameters you selected. By now, only one chromosome is evaluated in both workflows, but, downloading the OneMap or GUSMap object you can repeat the map building process in R for other chromosomes without spent all the time needed to compare all the methods. We show an example of this usage in [Quickstart]().


## Shiny app

This shiny app plots analyses datasets (.rds) generated by the Reads2Map. Upload your results in the `Upload Reads2Map workflow results` section. 

Each icon at the left corner of this page points to a different dataset and produces interactive plots. Also, you can download de graphics using the button at the end of the page.

## Documentation (under development)

Here are some tutorials that better explain how to use the workflows:

* [Introduction](https://cristianetaniguti.github.io/Reads2Map/introduction.html)
* [Running SimulatedReads2Map workflow](https://cristianetaniguti.github.io/Reads2Map/SimulatedReads2Map.html)
* [Running EmpiricalReads2Map workflow](https://cristianetaniguti.github.io/Reads2Map/EmpiricalReads2Map.html)
* [Configurations](https://cristianetaniguti.github.io/Reads2Map/configuration.html)
* [Usage of sub-workflows](https://cristianetaniguti.github.io/Reads2Map/subworkflows.html)

You can also have more details about the workflows and how they can be applied:

* [Paper]()

## Authors

Cristiane Taniguti is currently a PhD student at ESALQ/USP Statistical Genetics Lab. Access her personal page [here](http://cristianetaniguti.github.io/), and the page of the lab [here](http://augustogarcia.me/statgen-esalq/).
 