---
title: "About"
output: html_document
---

<!--
## OneMap software

The OneMap software was first developed by Gabriel Margarido, Marcelo Mollinari, and Antonio Augusto Franco Garcia and [published in 2007](http://taurus.unicamp.br/bitstream/REPOSIP/81752/1/WOS000248257800002.pdf). It was one of the first software able to build genetic maps for outcrossing species. It estimates a two-point recombination fraction for each combination of markers and applies a multipoint approach to combine information from every marker in the linkage group to estimate the genetic distances. But, the software was first developed to deal with few and informative genetic markers, as microsatellites. With the modern high-throughput markers, OneMap now has to deal with thousands of markers, with low-informativeness and a lot of genotyping errors. Without an adaptation of the software, it produces inflated maps and takes a lot of time to run.
-->

## Reads2Map workflows

[Reads2Map workflow](https://github.com/Cristianetaniguti/Reads2Map) uses [WDL language and cromwell](https://gatkforums.broadinstitute.org/wdl/) to build linkage maps from read sequences with optimized usage of memory, CPU, and time. There are two main workflows called SimulatedReads2Map and EmpiricalReads2Map. 

**SimulatedReads2Map** performs:

* Simulation of random polymorphisms or based on previous VCF file; 
* Simulation of mapping populations with PedigreeSim software;
* Simulation of sequencing library (RADseq, exome or WGS);
* Alignment of reads to the reference chromosome with BWA MEM;
* SNP calling with GATK and freebayes;
* Genotype calling with GATK, freebayes, SuperMASSA, polyRAD and updog using as read count source the VCFs from the SNP calling and the alignment file (BAM);
* Linkage map building with OneMap and GUSMap

All analysis are made based on a reference chromosome sequence.

**EmpiricalReads2Map** performs:

* Alignment of empirical sequence reads with BWA MEM;
* SNP calling with GATK and freebayes;
* VCF filtering by random criterias and for selected chromosome;
* Genotype calling by GATK, freebayes, SuperMASSA, polyRAD and updog (only one chromosome);
* Linkage map building with OneMap and GUSMap (only one chromosome)

The main two workflows are composed of sub-workflows that can also be useful for users, they are located in the `tasks` directory. The sub-workflows are composed of tasks that can be found at the end of each `.wdl` file, `utils.wdl`, and `utilsR.wdl`.

[Docker](https://www.docker.com/) images were built to each one of the tasks, then users need only to download the most recent version of cromwell, java, and docker. If running without sudo permission users can use [singularity](https://singularity.lbl.gov/docs-docker) instead of directly docker. See `.configurations` directory and its [instructions](vignettes/under_development.html) to use singularity.

The main goals of Reads2Map workflows are to test software capabilities to build linkage maps in simulated scenarios (SimulatedReads2Map) and in empirical data (EmpiricalReads2Map). SimulatedReads2Map also gives a safe method to test new functionalities implemented. The EmpiricalReads2Map gives a fast and automatic method to select the best software to build linkage maps.

## Reads2Map App

This shiny app plots graphics for descriptive analyses of the datasets generated by the Reads2Map workflows. Upload your results in the `Upload data` session. In this session you can also access previous obtained results.

Each icon at the left corner of this page points to a different dataset and produces interactive plots.

In the session "Maps" it is possible to download the OneMap or GUSMap R object containing the combination of software and parameters you selected. With the object you can make modifications in the sequence using the R environment and OneMap or GUSMap packages functions. 

By now, only one chromosome is evaluated in both workflows. However, the EmpiricalReads2Map workflow also provide the complete VCF file to users reproduce only the chosen pipeline in the other chromosomes. Therefore, users can select the best pipeline using one chromosome and apply it to the others without spent all the time and computer resources needed to compare all the methods for all chromosomes. We showed an example of this usage in [Quickstart](vignettes/under_development.html)

### Running the Reads2Map App locally

**warning**: The repositories are still private, we will publish them together with the papers.

To not overload this server, we limited the upload size to 500Mb by a user, but if you have larger result files you can run the App locally via GitHub using:

```{r, eval=FALSE}
devtools::install_github('Cristianetaniguti/Reads2MapApp')
Reads2MapApp::run_app()
```

It will also install all R packages dependencies needed for the App. If you do not want to install anything on your computer, you can access the App running the docker container:

```{bash}
docker run -p 8085:80  cristaniguti/reads2mapapp
```

The App will be available in `localhost:8085` or `127.0.0.1:8085`.

## Documentation (under development)

Here are some tutorials that better explain how to use the workflows:

* [Introduction](vignettes/under_development.html)
* [Running SimulatedReads2Map workflow](vignettes/under_development.html)
* [Running EmpiricalReads2Map workflow](vignettes/under_development.html)
* [Workflows configurations](vignettes/under_development.html)
* [Usage of sub-workflows](vignettes/under_development.html)

You can also have more details about the workflows and how they can be applied:

* [Paper](vignettes/under_development.html)


<!--

## Authors

Cristiane Taniguti is currently a PhD student at ESALQ/USP Statistical Genetics Lab. Access her personal page [here](http://cristianetaniguti.github.io/), and the page of the lab [here](http://augustogarcia.me/statgen-esalq/).
-->